### PR TITLE
Add Thread commision support in chip-repl

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
CHIP-repl only provides wifi provisioning.

 #### Summary of Changes
Add thread provisioning support. Also includes support for 'blob' thread generation.

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

File: src/controller/python/chip-repl.py [hunk 1]
```diff
@@ -62,6 +62,29 @@ def main():
 if device.needsNetworkCredentials:
   device.ConnectToWifi("ssid", "password")
 
+######## Thread provisioning ########
```
Comment (Damian-Nordic) [inline src/controller/python/chip-repl.py:65]: Out of curiosity, how are these changes related to `chip-device-ctrl.py` and `ChipDeviceCtrl.py`? I'm a bit confused which controller solution we should put effort in as after not-fully-completed switch from the native chip-tool to the Python controller it seems that new API within the Python controller is emerging. Is the new API going to replace bindings from `ChipDeviceCtrl.py` at some point?

**Review by Damian-Nordic — APPROVED** [review]

File: src/controller/python/chip-repl.py (hunk 1 above)
Comment (andy31415) [reply src/controller/python/chip-repl.py:65]: I am trying to get a scriptable environment - I have found tha chip-device-ctrl relies on interactive use and I would like to aim for either interactive (ipython) or automated scripting use. 

My long term goal would be to replace/remove device-ctrl completely and only keep the repl part.

File: src/controller/python/chip/internal/CommissionerImpl.cpp [hunk 2]
```diff
@@ -156,22 +165,51 @@ class ScriptDevicePairingDelegate final : public chip::Controller::DevicePairing
...
 {
     chip::python::ChipMainThreadScheduleAndWait([&]() { gPairingDelegate.SetWifiCredentials(ssid, password); });
 }
 
+extern "C" CHIP_ERROR pychip_internal_PairingDelegate_SetThreadCredentials(const void * data, uint32_t length)
+{
+
+    // Openthread is OPAQUE by the spec, however current CHIP stack does not have any
+    // validation/support for opaque blobs. As a result, we try to do some
+    // pre-validation hre
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/internal/CommissionerImpl.cpp:185]:
```suggestion
    // pre-validation here
```

File: src/controller/python/chip/internal/thread.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    "ThreadExtendedPANId" / Bytes(8),
+    "ThreadMeshPrefix" / Bytes(8),
+    "ThreadMasterKey" / Bytes(16),
+    "ThreadPSKc" / Bytes(16),
+    "_ThreadPANIdPadding" / Padding(1), # Aligns ThreadPANId
+    "ThreadPANId" / Int16ul,
+    "ThreadChannel" / Byte,
+    "_ThreadDatasetTimestampPadding" / Padding(3), # Aligns ThreadDAtasetTimestamp
+    "ThreadDatasetTimestamp" / Int64ul,
+)
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/internal/thread.py:61]: Add the trailing newline?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/python/chip/internal/CommissionerImpl.cpp [hunk 4]
```diff
@@ -156,22 +165,51 @@ class ScriptDevicePairingDelegate final : public chip::Controller::DevicePairing
...
-extern "C" void pychip_internal_PairingDelegate_SetNetworkCredentials(const char * ssid, const char * password)
+extern "C" void pychip_internal_PairingDelegate_SetWifiCredentials(const char * ssid, const char * password)
 {
     chip::python::ChipMainThreadScheduleAndWait([&]() { gPairingDelegate.SetWifiCredentials(ssid, password); });
 }
 
+extern "C" CHIP_ERROR pychip_internal_PairingDelegate_SetThreadCredentials(const void * data, uint32_t length)
+{
+
+    // Openthread is OPAQUE by the spec, however current CHIP stack does not have any
```
Comment (mspang) [inline src/controller/python/chip/internal/CommissionerImpl.cpp:183]: This format is not the specified format. They have roughly the same contents, but this isn't the representation of that data that we're required to use.

We're not meant to do any validation of the dataset.

Comment (andy31415) [reply src/controller/python/chip/internal/CommissionerImpl.cpp:183]: Are you suggesting to remove this check completely?

My plan was to have it as a TODO to remove once our entire stack treats it as opaque. Until then, our stack requires a specific structure usage, so we have no choice but to memcpy the data (or cast content), which seems to likely just crash on invalid data. Figure a basic sanity check is easier to debug than a crash.

Totally agree on long term - this should move around opaque data and not interpret it.

**Review by jmartinez-silabs — APPROVED** [review]
